### ISSUE TITLE
Google Analytics SDK retireing on October 2019

### ISSUE THREAD

**Issue by basvdijk** [danwilson/google-analytics-plugin#550]
Recently Google announced it wil retire the Google Analytics SDK on October 2019. They state you can use Firebase Analytics instead. However Google Analytics show way more detail and realtime data over Firebase Analytics.

Are there any plans to make this plugin act as if it is a website so the hits and events still can be used in Google Analytics?

**Comment by victorsosa** [comment]
Yes you can still use it as a website and works.

**Comment by boynet** [comment]
> Yes you can still use it as a website and works.

is there more info about it? how do I use it as website?

**Comment by victorsosa** [comment]
> is there more info about it? how do I use it as website?
Just setup the Google Analytics account as website and use this component as always; as documented.

**Comment by spinninghamster** [comment]
Do you have a link to the announcement?

**Comment by basvdijk** [comment]
@victorsosa I've tried this, but unfortunately the `trackView` doesn't show up as a visited page. For example in the realtime view, you can't see which view the user visited. 

@danwilson Maybe there is a way to make a `trackPage` method which you can give an url instead of a view name?

@spinninghamster sure, https://support.google.com/analytics/answer/9167112?hl=en

**Comment by remisture** [comment]
@victorsosa Will existing apps continue to work without any changes?

**Comment by victorsosa** [comment]
@remisture yes, Just setup the Google Analytics account as website and use this component as always; as documented.

**Comment by remisture** [comment]
I have an existing GA account I have used with my app for several years. I can't seem to find a way to convert this account into website? I dont want to create a new account and deploy a new version of my app unless I really have to.

**Comment by victorsosa** [comment]
@remisture yes you have to create a new GA account.

**Comment by jaxtheking** [comment]
I am using this plugin with Ionic, where the app runs on a local server. If I was to create a new GA account for regular website, what am I to enter as the 'website URL'?

I'm trying to avoid switching to one of the Firebase plugins, so your advice would be greatly appreciated.

**Comment by simonbrazell** [comment]
As @basvdijk mentioned setting the GA project up as a 'Website' instead of an app doesn't seem to work correctly with `trackView`, the page loads are coming through with a title of '(not set)' instead of whatever is set in the `trackView` call:

![screen shot 2019-01-31 at 1 39 52 pm](https://user-images.githubusercontent.com/16571951/52027041-15898b80-255e-11e9-91c7-ba600943ed05.png)

**Comment by basvdijk** [comment]
Any updates on how to transfer the `trackView` into a track page?

**Comment by MichielJ87** [comment]
I am not sure if there is any news on this topic? :)
Kind regards,
Michiel

**Comment by JayBizzle** [comment]
Anyone successfully migrated the Firebase?

**Comment by basvdijk** [comment]
> Anyone successfully migrated the Firebase?
Yes, however Firebase shows very sparse info compared to analytics

**Comment by JayBizzle** [comment]
@basvdijk any other alternatives?

**Comment by basvdijk** [comment]
@JayBizzle I am looking into a server solution, where app statistic calls are proxied to Google Analytics and mimic it to be a website.

**Comment by JayBizzle** [comment]
@basvdijk sounds good. If you get started with this, let us know the repo, would love to chip in and help where possible 👍

**Comment by basvdijk** [comment]
@JayBizzle I'll post the repo link here when I have something :)

**Comment by basvdijk** [comment]
@JayBizzle You can view my first attempt at https://github.com/basvdijk/server-side-google-analytics-proxy Logging pageviews is already working

**Comment by boynet** [comment]
@basvdijk in that case why not just use google analytics for website? its gonna be the same thing but without the need for proxy server

**Comment by basvdijk** [comment]
@boynet in an Ionic application, Google Analytics for websites doesn't see page changes when navigating. When I imported the GA tag in my html I was unable to reach it from the Angular code of my Ionic app. This was the reason to use this Cordova plugin in the first place.

**Comment by boynet** [comment]
oh :/ never use ionic, cant you just run this code on page change?
`ga('send', 'pageview');`

**Comment by basvdijk** [comment]
> oh :/ never use ionic, cant you just run this code on page change?
> `ga('send', 'pageview');`

It more an Angular thing. ga is not easily available inside the services. Something like that. Will look into it and post a solution here if I find one :)

**Comment by basvdijk** [comment]
With the comment of @boynet about using the web Analytics I've spend quite some time and tried the following with no luck:
- ga.js
- gtag
- Calling Google Analytics measurement with Angular http library

Some observations:
- Calling Google Analytics measurement with Angular http library worked in the browser, not in a Cordova container. The console shows that doing post is not allowed
- ga.js and gtag had the is issue that their functions were not defined when I tried to call them. It looks like here is a scoping issue with Angular or something
- It looks like Cordova is kind of sandboxing stuff, and that Google sees the calls are being made from  `file://` as I could see in the Chrome inspector. Perhaps this makes it block the post call.
- When diving into this keep in mind that you test that you are able to set the title / path of a page. Sometimes I have a call working but GA showed them all as `/` for the route. So every page showed the same title.

**Comment by boynet** [comment]
it seems like there is settings to make it work in cordvoa https://stackoverflow.com/questions/11026916/how-to-use-google-analytics-with-phonegap-without-a-plugin

> // THIS IS FOR LOCALSTORAGE
> var GA_LOCAL_STORAGE_KEY = 'ga:clientId';
> // THIS IS FOR FILE URL SUPPORT
> ga('set', 'checkProtocolTask', function(){ /* noop */});

about the page title and url, it make sense that in single page application you would need to manually send the title and url like documented here:
https://developers.google.com/analytics/devguides/collection/analyticsjs/pages

**Comment by basvdijk** [comment]
@boynet I've found the same resources and tried them all. I could get the `ga.js` to work with the main page only. Child pages didn't have access to the `ga` function.

**Comment by basvdijk** [comment]
Ok, I've spend about a week getting this worked out. And you'll laugh your *ss off when you hear what was causing my issue with implementing the web version of GA.

TL;DR Check if you use an DNS adblocker

I added the ga.js code to my index.html on the desktop everything fine, on my mobile however the console showed it couldn't load ga.js. Tried gtag.js same issue. So I thought maybe a Cordova issue with not being allowed to grab data from the Google domain. Still nothing worked.

One night I opened my news app and saw a big white square in the middle of the screen where an add was supposed to be. Then I remembered I have set up the adguard.com DNS for ad filtering.

But what if it blocked tracking as well? So I disabled the custom DNS and there you go, Google Analytics scripts loaded without any issues 😂

**Comment by JayBizzle** [comment]
@basvdijk thanks for the update! Glad you got it sorted.

I'm just about to start updating one of our apps with the `ga.js` method. Just wondering if there is anything I may need to know or should it be pretty much plug-and-play?

Thanks

**Comment by remisture** [comment]
@JayBizzle Why not just migrate to Firebase Analytics? Google is deprecating GA for native for a reason.

**Comment by JayBizzle** [comment]
@remisture mainly because I've heard that Firebase just isn't as good and doesn't provide data in a decent way.

Willing to be proven wrong though. All we need to do is track `Screen Views` and `Events`

**Comment by basvdijk** [comment]
My list of reasons not to go for FA:
- Depends on Cordova which had broken our builds a couple of times
- GA.js will work independent of the framework
- FA doesn't report as easy as GA e.g. dashboards
- FA doesn't cover the features of GA
- GA.js is handy for PWA apps as well
- FA realtime visitor overview is not as clean as for GA.js
- GA.js is not depending on a third party, of which you need to hope they update the library when it is broken.

**Comment by JayBizzle** [comment]
@basvdijk any negatives you have noticed?

**Comment by basvdijk** [comment]
@JayBizzle GA.js doesn't support app versions logging. But further I didn't ran into any issues.

**Comment by JayBizzle** [comment]
Does the realtime view work for you. I have got GA installed now and everything looks like it should be working but I'm not seeing anything in the realtime view.

**Comment by basvdijk** [comment]
@JayBizzle It does, it was one of my requirements :)

**Comment by JayBizzle** [comment]
Yep, just got it work now.

Got everything switched now and everything looks good.

Time to deploy to production 😰

**Comment by malwatte** [comment]
How do you guys use AnalyticsJS? It is using
```
<script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-XXXXX-Y', 'auto');
ga('send', 'pageview');
</script>
```
?

**Comment by tscislo** [comment]
For me it turned out that this plugin is no longer needed. 
I have Angular 8 app wrapped in Cordova (so similar use case as with Ionic)
I registered new web site in Google Analytics. I modified code form GA to look like this:

`       
        const gaScript = document.createElement('script');
        gaScript.src = "https://www.google-analytics.com/analytics.js";
        gaScript.type = "text/javascript";
        document.getElementsByTagName('head')[0].appendChild(gaScript);
        gaScript.onload = (ev: Event) => {
            const GA_LOCAL_STORAGE_KEY = 'ga:clientId';
            window['ga']('create', 'XXXX-XXXXX', {
                'storage': 'none',
                'clientId': localStorage.getItem(GA_LOCAL_STORAGE_KEY)
            });
            window['ga'](function(tracker) {
                localStorage.setItem(GA_LOCAL_STORAGE_KEY, tracker.get('clientId'));
            });

            // THIS IS FOR FILE URL SUPPORT
            window['ga']('set', 'checkProtocolTask', function(){ /* noop */});
            console.log('GA OK!')
        }`

And then I can use GA in regular way:
window['ga']('send', 'pageview', screenName);
This is because there is no automatic tracking of page changes while serving using file:// protocol.

This solution is very handy as it is DRY, because I have just one way to embed Google Analytics in my app for both hybrid and hosted versions.

**Comment by jondspa** [comment]
I've been using this plugin for a few years, collecting analytics on a couple of Cordova apps.  It sounds like when I transition to Firebase, my old data goes away and I start from scratch.  Can this be true?  I certainly want to compare this month to last month, this year to last year, etc.

Please tell me it's not true!

- Jon

**Comment by tscislo** [comment]
@jondspa My understanding is (please correct me if I'm wrong) that you have two options going forward after October 2019:
1. Use Google Analytics for web sites and then no plugin is needed just look at my previous comment.
2. Go with Firebase Analytics integration
Unfortunately in both cases you loose your old data reported under old tracking ID.

**Comment by jondspa** [comment]
Thanks @tscislo!  This is so hard to believe that our Google friends didn't set up a clean transition path!  (I'm not using angular or ionic so not sure your first suggestion would work.  But as you say my old data goes away anyway.)

I have a friend who works at Google.  I wonder if I can ask him to "walk down the hall" to the Firebase people and see if there's a solution...

**Comment by tscislo** [comment]
@jondspa great! let us know!

**Comment by MichielJ87** [comment]
Hi all,

I was able to successfully migrate from the plugin to the 'normal' ga.js, using vanilla javascript.

For anyone trying this, I have some tips **tips**. Although they might be obvious to some of you, who knows it might help somebody running into the same problems like I did :)
- Make sure your code is no longer using any functions from the plugin
- Make sure to remove the plugin from your config.xml, otherwise the 'ga' function will not work
- When do your setup in Google Analytics, the suggested code for your property is using gtag.js, and not ga.js. I have used ga.js (see code snippet below) to be consistent with examples given in this thread
- Instead of loading in analytics.js from the web, I put it analytics.js in a local file. This way my app will still load when offline (and will not load slower on slow internet connections)

I also have one **question**:
How are you guys handling versioning in your analytics (to easily identify which version somebody has in google analtics reporting)?

For somebody trying to migrate from the plugin to vanilla JS, my code can be found below.

Kind regards,
Michiel


  code put in index.html head:
  <script>

    // tracking script from Google (using ga.js, and not gtag.js like mentioned in Google Analytics setup...)
    (function (i, s, o, g, r, a, m) {
      i['GoogleAnalyticsObject'] = r; i[r] = i[r] || function () {
        (i[r].q = i[r].q || []).push(arguments)
      }, i[r].l = 1 * new Date(); a = s.createElement(o),
        m = s.getElementsByTagName(o)[0]; a.async = 1; a.src = g; m.parentNode.insertBefore(a, m)
    })(window, document, 'script', 'js/analytics.js', 'ga'); /* local copy from web 30/08/19 https://www.google-analytics.com/analytics.js */

    // THIS IS FOR LOCALSTORAGE (info from https://stackoverflow.com/questions/11026916/how-to-use-google-analytics-with-phonegap-without-a-plugin)
    var GA_LOCAL_STORAGE_KEY = 'ga:clientId';
    ga('create', 'UA-XXXXXXX-Y', {
      'storage': 'none',
      'clientId': localStorage.getItem(GA_LOCAL_STORAGE_KEY)
    });
    ga(function (tracker) {
      localStorage.setItem(GA_LOCAL_STORAGE_KEY, tracker.get('clientId'));
    });

    // THIS IS FOR FILE URL SUPPORT (info from https://stackoverflow.com/questions/11026916/how-to-use-google-analytics-with-phonegap-without-a-plugin)
    ga('set', 'checkProtocolTask', function () { /* noop */ });

  </script>

**Comment by basvdijk** [comment]
> When do your setup in Google Analytics, the suggested code for your property is using gtag.js, and not ga.js. I have used ga.js (see code snippet below) to be consistent with examples given in this thread

When you do this keep in mind that you have to do a new app release when Google updates the analytics.js with breaking changes.

**Comment by malwatte** [comment]
@MichielJ87 
Sending GA hits will fail if there is no internet even if the library is loaded. Keeping the GA library locally is not recommend by Google. So what I did was to load the library from Google Analytics whenever internet is available.
I use ga, not gtag. Should I use gtag?

**Comment by monkeytronics** [comment]
@tscislo @MichielJ87 @basvdijk Amazing work! So to add google analytics to an ionic app, you advise not using the plugin but game as above? Should be able to generate analytics breakdown by ios; android; web this way?

**Comment by basvdijk** [comment]
@monkeytronics I don't use the plugin anymore. I directly init Google Analyics in my index.html

```
  <script>
    // https://stackoverflow.com/questions/11026916/how-to-use-google-analytics-with-phonegap-without-a-plugin
    // the default GA code, nothing to change
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    var fields = {
      // note: you can use a single tracking id for both the app and the website,
      // don't worry it won't mix the data. More about this in the 3rd section
      trackingId: 'UA-XXXXXXXX-X'
    };

    // we store and provide the clientId ourselves in localstorage since there are no
    // cookies in Cordova
    fields.clientId = localStorage.getItem('ga:clientId');
    // disable GA's cookie storage functions
    fields.storage = 'none';

    ga('create', fields);
    ga('set', 'anonymizeIp', true);

    // prevent tasks that would abort tracking
    ga('set', {
      // don't abort if the protocol is not http(s)
      checkProtocolTask: null,
      // don't expect cookies to be enabled
      checkStorageTask: null
    });

    // a callback function to get the clientId and store it ourselves
    ga(function(tracker){
      localStorage.setItem('ga:clientId', tracker.get('clientId'));
    });

  </script>
```

Then I've create a statistics service:

```
import { Injectable } from '@angular/core';

// CORE
declare var ga: Function;

/**
 * 
 * Service used for tracking user events or page views
 * https://developers.google.com/analytics/devguides/collection/analyticsjs/pages
 * @export
 * @class StatisticsService
 */
@Injectable()
export class StatisticsService {

    constructor() { }

    /**
     * Track a page view 
     * @param {string} viewPath Title of the view
     * @returns {Promise<any>}
     * 
     * @memberOf StatisticsService
     */
    trackView(viewPath: string, viewTitle?: string) {

        ga('send', {
            hitType: 'pageview',
            title: viewTitle,
            location: '/' + viewPath
        });

    }
}
```

Which I use to track all the views.

**Comment by monkeytronics** [comment]
Thanks @basvdijk. I've added that code into index.html.

Can't see anything right away. I thought I might see the initial page load. Maybe there is a delay til it's active. In GA, I assume you select 'Web'. Does it matter what URL you select. I choose the base url of the web app. 

My app is Ionic 4. It uses the Angular Router. Where are you calling the statistics.trackview() function from?

**Comment by basvdijk** [comment]
@monkeytronics You need to call `trackView` for every page manually (I am using ionic 3) With Angular router you can write a route interceptor which calls `trackView` on every route change.

**Comment by tscislo** [comment]
@monkeytronics here is a service that I made to track my page views:
`
import {Injectable} from '@angular/core';
import {addScriptTag} from '../shared/utils';
import {BehaviorSubject} from "rxjs";

@Injectable()
export class AnalyticsService {

    public isInitialized$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);

    private tracker;
    private originalSendHitTask;

    constructor() {
        addScriptTag('https://www.google-analytics.com/analytics.js').onload = () => {
            const ga = window['ga'];

            const GA_LOCAL_STORAGE_KEY = 'ga:clientId';
            ga('create', 'GA_KEY', {
                'storage': 'none',
                'clientId': localStorage.getItem(GA_LOCAL_STORAGE_KEY)
            });
            ga((tracker) => {
                this.originalSendHitTask = tracker.get('sendHitTask');
                this.tracker = tracker;
                this.enableDisableTracking(false);
                localStorage.setItem(GA_LOCAL_STORAGE_KEY, tracker.get('clientId'));
            });

            // THIS IS FOR FILE URL SUPPORT
            ga('set', 'checkProtocolTask', function () { /* noop */
            });
            console.log('Google Analytics loaded!');
            this.isInitialized$.next(true);
        };
    }

    private enableDisableTracking(enable: boolean) {
        if (enable) {
            this.tracker.set('sendHitTask', this.originalSendHitTask);
        } else {
            this.tracker.set('sendHitTask', null);
        }
    }

    public trackPage(pageName: string) {
        try {
            // here we enable tracking
            this.enableDisableTracking(true);
            window['ga']('set', 'page', pageName);
            window['ga']('send', 'pageview');
            // here we disable auto tracking, because we want to specifically track pageviews for us
            this.enableDisableTracking(false);
            console.log('tracking', pageName);
        } catch (e) {
            console.warn('No GA...');
        }
    }
}
`
`
export function addScriptTag(url: string) {
    const script = document.createElement('script');
    script.src = url;
    script.type = 'text/javascript';
    document.getElementsByTagName('head')[0].appendChild(script);
    return script;
}

`
